feat: Add support for UUID comparison functions

BryanCutler opened the pull request:
This adds binary comparison functions <, >, <=, >= to the UUID custom data type. Equality functions are already present. Added unit tests for testing a query with comparisons between UUID values.

The ordering is done lexicographically to conforms with IETF RFC 4122 https://datatracker.ietf.org/doc/html/rfc4122.html, and also matches Presto Java after https://github.com/prestodb/presto/issues/23311.

Related UUID serialization fix at https://github.com/facebookincubator/velox/pull/11197
From https://github.com/facebookincubator/velox/issues/10584

aditi-pandit commented on velox/functions/prestosql/UuidFunctions.h:
```diff
@@ -20,10 +20,46 @@
+  }
+};
+
+#define VELOX_GEN_BINARY_EXPR_UUID(Name, uuidCompExpr, TResult)               \
```
TResult parameter is not used.

BryanCutler replied:
Thanks, good catch! It's actually the same in `VELOX_GEN_BINARY_EXPR_TIMESTAMP_WITH_TIME_ZONE` too

BryanCutler replied:
done

aditi-pandit reviewed (commented):
Thanks @BryanCutler. One small comment.

Did you try running the queries in https://github.com/prestodb/presto/pull/23301 with this logic ?

BryanCutler commented:
> Did you try running the queries in prestodb/presto#23301 with this logic ?

Yes, and the queries for <,>,<=,>= now pass. But it is a little strange, they only pass for me in debug mode, when I run in release mode I get this type of error:
```
not equal
Actual rows (1 of 1 extra rows shown, 6 rows in total):
    [f768f36d-4f09-4da7-a298-3564d8f3c986, 1fdff429-6f65-4370-bb2a-7ab660dbec92]
Expected rows (1 of 1 missing rows shown, 6 rows in total):
    [1fdff429-6f65-4370-bb2a-7ab660dbec92, f768f36d-4f09-4da7-a298-3564d8f3c986]
```
where it looks like one UUID is randomly generated, and I'm not sure where it's coming from. I'll look into that a bit more.

BryanCutler commented:
@aditi-pandit I didn't realize the temp table from the test in https://github.com/prestodb/presto/pull/23301 included a randomly generated UUID with `(cast(uuid() AS VARCHAR))`. That was the reason for failing in release mode, without this random value all the comparison tests pass.

aditi-pandit reviewed (commented):
Thanks @BryanCutler. Had some minor comments.

BryanCutler commented on velox/functions/prestosql/UuidFunctions.h:
```diff
@@ -20,10 +20,46 @@
+struct UuidCompareBase {
+  FOLLY_ALWAYS_INLINE int compare64(uint64_t x, uint64_t y)
+  {
+    return x < y ? -1 : (x == y ? 0 : 1);
```
No, just that the return values read -1, 0, 1 left to right

BryanCutler commented on velox/functions/prestosql/tests/UuidFunctionsTest.cpp:
```diff
@@ -108,5 +108,45 @@ TEST_F(UuidFunctionsTest, unsupportedCast) {
 
+TEST_F(UuidFunctionsTest, comparisons) {
+  auto data0 = makeFlatVector<std::string>({
+      "33355449-2c7d-43d7-967a-f53cd23215ad",
```
Sure, for the lower bits I have a check for <, = and I'll add one more for >

BryanCutler replied:
added additional testing for lower

BryanCutler commented on velox/functions/prestosql/UuidFunctions.h:
```diff
@@ -20,10 +20,46 @@
+  template <typename T>                                                       \
+  struct Name##Uuid : public UuidCompareBase {                                \
+    VELOX_DEFINE_FUNCTION_TYPES(T);                                           \
+    FOLLY_ALWAYS_INLINE void call(                                            \
```
@aditi-pandit I don't see a consistent format for this, most of the time it looks how it is currently but I do see some cases where the format is:
```
template <typename T>
struct MultiplyVoidOutputFunction {
  template <typename TInput>
  FOLLY_ALWAYS_INLINE void
  call(TInput& result, const TInput& a, const TInput& b) {
    result = functions::multiply(a, b);
  }
};
```
Do you mean the latter, with a newline after void?

aditi-pandit replied:
@BryanCutler : I meant newline in between 
 ```
VELOX_DEFINE_FUNCTION_TYPES(T);      
                                     \
    FOLLY_ALWAYS_INLINE void call(      
```

aditi-pandit reviewed (approved):
Thanks @BryanCutler.

aditi-pandit commented:
@Yuhta @kagamiori @mbasmanova : Please can you help with the review.

BryanCutler commented:
Gentle ping @Yuhta @kagamiori @mbasmanova to please review

mbasmanova commented on velox/functions/prestosql/UuidFunctions.h:
```diff
@@ -20,10 +20,47 @@
 namespace facebook::velox::functions {
 
+struct UuidCompareBase {
+  FOLLY_ALWAYS_INLINE int compare64(uint64_t x, uint64_t y)
```
These can be static functions or free functions.

Let's also put this code into 'details' namespace.

mbasmanova commented on velox/functions/prestosql/UuidFunctions.h:
```diff
@@ -20,10 +20,47 @@
+  }
+
+  int compare128(const int128_t& lhs, const int128_t& rhs) {
+    int comp = compare64(HugeInt::upper(lhs), HugeInt::upper(rhs));
```
> comp

Do not abbreviate: comp -> result or similar.

mbasmanova commented on velox/functions/prestosql/tests/UuidFunctionsTest.cpp:
```diff
@@ -108,5 +108,49 @@ TEST_F(UuidFunctionsTest, unsupportedCast) {
+      "33355449-2c7d-43d7-967a-f53cd23215ad",
+  });
+
+  auto actual = evaluate("cast(c0 as uuid) < cast(c1 as uuid)", makeRowVector({data0, data1}));
```
Any particular reason not to test one value at a time using evaluateOnce? This makes tests easier to read and debug.

BryanCutler replied:
No, just not familiar with the test environment. I will try with `evaluateOnce`.

mbasmanova reviewed (commented):
@BryanCutler Thank you for adding this functionality. I wonder where is the semantics of comparing UUIDs come from? Do you have a pointer? Would be nice to update PR description and documentation (https://facebookincubator.github.io/velox/develop/types.html) to clarify.

BryanCutler commented:
Thanks for reviewing @mbasmanova 

> I wonder where is the semantics of comparing UUIDs come from? Do you have a pointer? Would be nice to update PR description and documentation

I basically followed the comparison done in in Presto Java https://github.com/prestodb/presto/blob/97d88cfdebb553b5c7fb0e217b0211573df3d190/presto-common/src/main/java/com/facebook/presto/common/type/UuidType.java#L97

Since int128_t doesn't have support for atomic operations, we first check the higher bytes, then lower. Returning an int from the comparison allows it to be streamlined to do minimal checks. I will update the PR and documentation.

mbasmanova commented:
@BryanCutler I wonder whether it makes sense to compare UUIDs at all? What does it mean that one UUID is greater than the other? Perhaps, this type should be non-orderable.

BryanCutler commented:
> @BryanCutler I wonder whether it makes sense to compare UUIDs at all? What does it mean that one UUID is greater than the other? Perhaps, this type should be non-orderable.

@mbasmanova I think it makes sense, just as you would compare values for std::byte, etc. Most UUID implementations support such operations, see https://www.boost.org/doc/libs/1_85_0/libs/uuid/doc/uuid.html#Operators

aditi-pandit commented:
> > @BryanCutler I wonder whether it makes sense to compare UUIDs at all? What does it mean that one UUID is greater than the other? Perhaps, this type should be non-orderable.
> 
> @mbasmanova I think it makes sense, just as you would compare values for std::byte, etc. Most UUID implementations support such operations, see https://www.boost.org/doc/libs/1_85_0/libs/uuid/doc/uuid.html#Operators

@mbasmanova : I found UUID comparisons in the Java spec also https://docs.oracle.com/javase/8/docs/api/java/util/UUID.html

My understanding is that a source generating UUIDs uses a combination of timestamp with the DNS address of the generating device. So there is some notion of ordering of UUIDs as a result.

BryanCutler commented:
@mbasmanova just wanted to confirm if you are ok with adding this functionality? I still need to update the tests per your request and will get to that soon, thanks!

mbasmanova commented:
@BryanCutler Yes, please, move ahead with addressing the comments. Thank you for clarifying the use case.

BryanCutler commented:
@mbasmanova @aditi-pandit after looking into https://github.com/prestodb/presto/issues/23311 there are a couple issues with UUIDs in Velox:

1. UUID data is being stored as big endian in the int128_t value, which comes from using boost:uuid during casting to/from strings. This causes comparisons like `11000000-0000-0022-0000-000000000000 < 22000000-0000-0011-0000-000000000000` to fail.
2. Serialization of the int128_t value to an int128 Array Block results in [LSW, MSW], however the Presto Java UUIDType expects long values from the int128 Array Block to be [MSW, LSW].

For (1), since Velox stores UUID values with an int128_t, unlike boost::uuid which is a byte array, I think the value should be in little endian. This means instead of directly using memcpy bytes from the boost::uuid during casting, the bytes would need to swapped to get int128 in LE order. wdyt?

To fix (2), the word order would need to be switched during serialization. Would you prefer I fix that in this PR or open another?

aditi-pandit commented:
I feel the first fix should be to address is the issue of the incorrect results in https://github.com/prestodb/presto/issues/23311. Post that we can do this comparisons PR.

From your comments seems that requires 2 changes:
i) Fixing the seriailization order between the native and java sides (2) above
ii) Changing the uuid int128_t to be little endian in the copy from boost::uuid. 

@mbasmanova : Using boost::uuid (instead of int128_t) to represent UUID in Velox could be cleaner given these issues. We could make UUID TypeKind::VARBINARY and use the boost::uuid bytes to represent it. wdyt ?

@BryanCutler : If we did that, would we need the serialization changes from Native to Java side as well ? My intuition is yes, but I could be wrong.

BryanCutler commented:
Thanks @aditi-pandit , if we did change the storage from int128_t to a byte array in boost::uuid, then we would still need to adjust the serialization to Java, since Presto expects an Int128ArrayBlock.

Yuhta commented:
@aditi-pandit I don't think VARBINARY is the right physical type for UUID, it is meant for variable length data, while UUID is fixed length.  I would say just keep using int128 and fix the endianness so the comparison between UUID is the same as comparison between int128.

aditi-pandit commented:
Thanks @Yuhta for your response. I do see a pattern of using VARBINARY for fixed lengths (in IPPREFIX type) but that has length 17 bytes which can't be captured with any integral type. For UUID we do have the option of using int128_t.

@BryanCutler : Let's go with Jimmy's suggestion. Will make our implementation simpler as well.

BryanCutler commented:
Thanks @Yuhta and @aditi-pandit , I have fixed the serialization issue here and added a unit test, I can put that in a separate PR if preferred. I will also test these changes again with https://github.com/prestodb/presto/issues/23311

BryanCutler commented:
After testing the comparisons with Presto Java in https://github.com/prestodb/presto/issues/23311 there are some different results because the Java UUID `compareTo` uses signed values, so when the uuid bytes overflow it becomes negative and does not follow the behavior defined in [RFC-4122](https://datatracker.ietf.org/doc/html/rfc4122). This is a known issue in Java, see https://bugs.openjdk.org/browse/JDK-7025832 and we will have to discuss how to handle this.

In the meantime, I'll separate the serialization issue, which is more straight-forward and will at least start show the correct values in Presto Java.

Yuhta commented on velox/functions/prestosql/UuidFunctions.h:
```diff
@@ -20,10 +20,49 @@
+  return x < y ? -1 : (x == y ? 0 : 1);
+}
+
+static int compare128(const int128_t& lhs, const int128_t& rhs) {
```
Just compare `(uint128_t)lhs` and `(uint128_t)rhs`

BryanCutler commented:
> Just compare (uint128_t)lhs and (uint128_t)rhs

Thanks @Yuhta , I can look into that but there is a bigger concern that this type of ordering will not match Presto Java, see above https://github.com/facebookincubator/velox/pull/10791#issuecomment-2400723075.

Presto Java `UuidOperations` use `Slice.compareTo()` which follows a strange ordering based on swapping bytes. As far as I can tell this ordering was changed in the Slice library 2.0+, which uses `Arrays.compareUnsigned` and I don't believe it would have the same problem, but not sure upgrading that is planned in the near-term.

I'm not sure if we should try to follow Presto Java with the comparisons or use the spec at [RFC-4122](https://datatracker.ietf.org/doc/html/rfc4122) to order lexicographically, e.g. comparing uint128_t in little endian, and differ from Presto Java.  Please share recommendations on this @Yuhta @aditi-pandit

BryanCutler commented:
There is a recent related PR cherry-picked from trindodb, this might address some part of the Presto Java issue https://github.com/prestodb/presto/pull/17939.

Yuhta commented:
@BryanCutler We should ask Presto Java community to decide the ordering of UUID and document them officially first.

BryanCutler commented:
> @BryanCutler We should ask Presto Java community to decide the ordering of UUID and document them officially first.

There is a PR up at https://github.com/prestodb/presto/pull/23847 that corrects Presto Java to conform to the IETF RFC 4122 definition of ordering UUIDs. When that is resolved, this PR can be updated to match.

BryanCutler commented:
Status: PR https://github.com/prestodb/presto/pull/23847 has been merged, and I have updated the comparisons to use a cast to uint128_t, and this includes the serialization commit from https://github.com/facebookincubator/velox/pull/10791 which stores the UUID value as little endian. Once that PR is merged, it can be dropped from here. 

I have also checked these changes again to ensure the Presto Java tests from https://github.com/prestodb/presto/pull/23301

BryanCutler commented:
Finished rebasing after fixing serialzation issue at https://github.com/facebookincubator/velox/pull/10791

aditi-pandit commented on velox/functions/prestosql/tests/UuidFunctionsTest.cpp:
```diff
@@ -108,5 +108,104 @@ TEST_F(UuidFunctionsTest, unsupportedCast) {
+          "33355449-2c7d-43d7-967a-f53cd23215ad",
+          "<",
+          "ffffffff-ffff-ffff-ffff-ffffffffffff"));
+  ASSERT_EQ(
```
It would be great to have more tests where the lower characters are different.

Could we use some pattern to construct 2 UUIDs such that first 2 bytes of 1st < first 2 bytes of second, 2nd 2 bytes of 1st > 2nd 2 bytes of 2nd, 3rd 2 bytes of 1st < 3rd 2 bytes of 2nd something like :

12342345-3456-4567-5678-678978908901
23451234-4567-3456-6789-567889017890

and then do comparisons by making the first set of bytes equal as you go
So the first comparison would be 
12342345-3456-4567-5678-678978908901 < 23451234-4567-3456-6789-567889017890, 

but the 2nd case would be
00002345-3456-4567-5678-678978908901 > 00001234-4567-3456-6789-567889017890,

 and the 3rd would be 00000000-3456-4567-5678-678978908901< 00000000-4567-3456-6789-567889017890

BryanCutler replied:
testing against FFFF.... is necessary to make sure it handles the integer overflow property. I do do test a few targeted values where lower portions are different but I can also setup a loop to test what you're describing if you think there should be more.

BryanCutler replied:
ok added, lmkwyt

aditi-pandit replied:
Thanks !

BryanCutler commented on velox/functions/prestosql/tests/UuidFunctionsTest.cpp:
```diff
@@ -108,5 +108,104 @@ TEST_F(UuidFunctionsTest, unsupportedCast) {
+  const auto uuidEval = [&](const std::optional<std::string>& lhs,
+                            const std::string& operation,
+                            const std::optional<std::string>& rhs) {
+    return evaluateOnce<bool>(
```
That does work, but then I would need to construct the UUIDs with boost and then convert to an int128_t to pass to the eval. I think the end result is less clear, and this way also tests the UUID casting.

aditi-pandit reviewed (approved):
Thanks @BryanCutler for working with the revisions.

aditi-pandit commented:
@Yuhta : Please can you help further review and merge.

Yuhta commented:
@BryanCutler Can you make sure the failed tests is related or not?
https://github.com/facebookincubator/velox/actions/runs/12128994244/job/33818706033?fbclid=IwZXh0bgNhZW0CMTEAAR13mXks2oVNMArutfUoHnSSEWNU9OqKgAzKSe8NEcDoWjHe8BNZUhElEXc_aem_DC2AdtBRx0Wvfd-JJvU_HQ

BryanCutler commented:
It seems like one of the problems is the biased expression fuzzer test tries to create a hive table with IPADDRESS or UUID as a type, and they are not supported so the reference query fails. 

Locally, this is one of the errors I see
```
I20241210 12:16:31.433363 354597 PrestoQueryRunner.cpp:804] Execute presto sql: CREATE TABLE tmp(c0, c1, row_number) WITH (format = 'DWRF') AS SELECT cast(null as IPADDRESS), cast(null as BOOLEAN), cast(null as BIGINT)
E20241210 12:16:31.447225 354597 Exceptions.h:66] Line: /home/bryan/git/velox/velox/exec/fuzzer/PrestoQueryRunner.cpp:98, Function:throwIfFailed, Expression:  Presto query failed: 65536 Unsupported type: ipaddress, Source: RUNTIME, ErrorCode: INVALID_STATE
```

The test might have to be adjusted for these types, I'll look into what we can do about it.

BryanCutler commented:
Another PR has failed biased expression test, not quite the same error since testing different functions https://github.com/facebookincubator/velox/actions/runs/12279548356/job/34266985156?pr=11612

xiaoxmeng commented:
@BryanCutler can you rebase and make sure all tests green? Thanks!

BryanCutler commented:
The Spark fuzzer failure looks related to https://github.com/facebookincubator/velox/pull/12000

BryanCutler commented:
@rui-mo after rebasing I still got the following error from Spark Fuzzer test
```
E20250108 22:59:35.849556    40 Exceptions.h:66] Line: /__w/velox/velox/velox/velox/type/tz/TimeZoneMap.cpp:251, Function:validateRangeImpl, Expression:  Timepoint is outside of supported year range: [-32767, 32767], got 39215, Source: RUNTIME, ErrorCode: UNSUPPORTED_INPUT_UNCATCHABLE
```
Any ideas if this is different from your fix?

rui-mo commented:
Hi @BryanCutler, I assume it is different. It looks more like the case mentioned in https://github.com/facebookincubator/velox/issues/11462#issuecomment-2463937695, where the Spark fuzzer test times out after 2 hours and got canceled.

BryanCutler commented:
@xiaoxmeng looks like all tests are passing now
